File: react-frontend/src/services/taskDataSourceService.ts
import api from '../api';
import { DataSource } from '../models/DataSource';
import { StrapiServiceResponse } from '../types/StrapiServiceResponse';
import { ErrorResponse } from '../types/ErrorResponse';
import { Tag } from '../models/Tag';
import { TaskDataSource } from '../models/TaskDataSource';

const taskDataSourceService = {
  getAll: async (keycloakSubject: string | undefined): Promise<StrapiServiceResponse<TaskDataSource[]>> => {
    const currentDate = new Date().toISOString();
    try {
      if (!keycloakSubject) {
        const result: StrapiServiceResponse<TaskDataSource[]> = {
          data: null,
          error: {
            status: 400,
            statusText: 'Bad Request',
            data: { message: 'subject is undefined' },
          },
        };
        return result;
      }
      const response = await api.get(`/task-data-sources?filters[userProfile][keycloaksubject][$eq]=${keycloakSubject}`);
     
      const result: StrapiServiceResponse<TaskDataSource[]> = { data: response.data.data, error: null };
      return result;
    } catch (error) {
      const result: StrapiServiceResponse<TaskDataSource[]> = { data: null, error: error as ErrorResponse };
      return result;
    }
  },
  getByTaskId: async (id: string | undefined): Promise<StrapiServiceResponse<TaskDataSource[]>> => {
    try {

      if (!id) {
        const result: StrapiServiceResponse<TaskDataSource[]> = {
          data: null,
          error: {
            status: 400,
            statusText: 'Bad Request',
            data: { message: 'Task ID is undefined' },
          },
        };
        return result;
      }

      const response = await api.get(`/task-data-sources/?populate[0]=dataSource&populate[1]=task$filters[task][id][$eq]=${id}`);
      const result: StrapiServiceResponse<TaskDataSource[]> = { data: response.data.data, error: null };
      return result;
    } catch (error) {
      const result: StrapiServiceResponse<TaskDataSource[]> = { data: null, error: error as ErrorResponse };
      return result;
    }
  },
  create: async (taskDataSource: TaskDataSource): Promise<StrapiServiceResponse<TaskDataSource>> => {
    try {
      const response = await api.post('/task-data-sources?populate[0]=task&populate[1]=dataSource', { data: taskDataSource });
      const result: StrapiServiceResponse<TaskDataSource> = { data: response.data.data, error: null };
      return result;
    } catch (error) {
      const result: StrapiServiceResponse<TaskDataSource> = { data: null, error: error as ErrorResponse };
      return result;
    }
  },
  delete: async (id: number | undefined): Promise<StrapiServiceResponse<TaskDataSource>> => {
    try {
      await api.delete(`/task-data-sources/${id}`);
      return { data: null, error: null };
    } catch (error) {
      return { data: null, error: error as ErrorResponse };
    }
  },
};

export default taskDataSourceService;

File: vector-ask/src/main/java/com/enterprisewebservice/api/TaskDataSourceService.java
package com.enterprisewebservice.api;

import com.enterprisewebservice.model.ErrorResponse;
import com.enterprisewebservice.model.Note;
import com.enterprisewebservice.model.TaskNote;
import com.enterprisewebservice.model.datasource.TaskDataSource;
import com.enterprisewebservice.model.StrapiServiceResponse;
import com.fasterxml.jackson.databind.ObjectMapper;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import java.io.IOException;
import java.util.List;

@ApplicationScoped
public class TaskDataSourceService {

    @Inject
    ApiClient apiClient;

    private final ObjectMapper objectMapper = new ObjectMapper(); // replace with your actual JSON library


    public StrapiServiceResponse<List<TaskDataSource>> getByTaskId(String id) throws IOException {
        if (id == null || id.isEmpty()) {
            return new StrapiServiceResponse<>(null, new ErrorResponse(400, "Bad Request", "Task ID is undefined"));
        }

        String response = null;
        try {
            response = apiClient.get("/task-data-sources?populate[0]=task&populate[1]=dataSource&filters[task][id][$eq]=" + id);
        } catch (Exception e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
        return objectMapper.readValue(response, objectMapper.getTypeFactory().constructParametricType(StrapiServiceResponse.class, objectMapper.getTypeFactory().constructCollectionType(List.class, TaskDataSource.class)));
    }
   
}
